### PR TITLE
Versions in extensions

### PR DESCRIPTION
**Pull request by ekr**

### CODE REVIEW

File: draft-ietf-tls-tls13.md [hunk 1]
```diff
@@ -1412,11 +1412,16 @@ before extensions were defined.
 As of TLS 1.3, all clients and servers will send at least
 one extension (at least "key_share" or "pre_shared_key").
 
-max_supported_version
-: The latest (highest valued) version of the TLS protocol offered by the
-  client. This SHOULD be the same as the latest version supported. For this
-  version of the specification, the version will be { 3, 4 }. (See
-  {{backward-compatibility}} for details about backward compatibility.)
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: The back compat ref should probably stay on here.

File: draft-ietf-tls-tls13.md [hunk 2]
```diff
@@ -1626,6 +1631,7 @@ The extension format is:
            pre_shared_key(41),
            early_data(42),
            cookie(44),
+           supported_versions(46),
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: 43 is free; can we just stick it there to appease nit issue #534?

File: draft-ietf-tls-tls13.md [hunk 3]
```diff
@@ -4223,8 +4262,9 @@ whenever TLS 1.3 is used.
 ## Negotiating with an older server
 
 A TLS 1.3 client who wishes to negotiate with such older servers will send a
-normal TLS 1.3 ClientHello containing { 3, 4 } (TLS 1.3) in
-ClientHello.max_supported_version. If the server does not support this version it
+normal TLS 1.3 ClientHello containing { 3, 3 } (TLS 1.2) in
+ClientHello.max_supported_version but with the correct version in the
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: need to rename all use of max_supported_version to legacy_version

File: draft-ietf-tls-tls13.md [hunk 4]
```diff
@@ -1692,6 +1698,35 @@ be taken into account when designing new extensions:
...
+prepared to negotiate (for this specification, that means minimally
+{3, 4}, but if previous versions of TLS are supported, they MUST
+be present as well).
+
+Servers which are compliant with this specification MUST use
+only the "supported_versions" extension to determine client preferences
+and MUST only select a version of TLS present in that extension.
+They MUST ignore any unknown versions. If the extension is not present,
+they MUST negotiate TLS 1.2 or prior as specified in {{RFC5246}},
+even if ClientHello.version is {3, 4} or later.
```
Comment (xiaoyinl) [inline draft-ietf-tls-tls13.md]: Shouldn't that be "ClientHello.legacy_version"?

File: draft-ietf-tls-tls13.md [hunk 5]
```diff
@@ -1395,7 +1395,7 @@ Structure of this message:
        uint8 CipherSuite[2];    /* Cryptographic suite selector */
 
        struct {
-           ProtocolVersion max_supported_version = { 3, 4 };    /* TLS v1.3 */
+           ProtocolVersion legacy_version = { 3, 3 };    /* TLS v1.2 */
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: Why do we persist in using a struct for the version, rather than just changing to a uint16-based enum?

Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: I'd vote for going uint16, as I did in my version of this proposal. One way or another, we really should do a formal registry, as was suggested on-list in the TLS 1.3 vs 2.0 discussion. I'll submit a PR with just that cherry-picked out of the rebranding PR, as it's a good idea to make this easier to follow, regardless of what we call all this.

Comment (davidben) [inline draft-ietf-tls-tls13.md]: Registry and uint16 SGTM. A registry will be useful anyway to burn things like private use values for drafts and GREASE values if the WG is so inclined.

Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: I've pushed PR #634 with just the changeset for a registry and explicit naming of the values as IDs.

Comment (richsalz) [inline draft-ietf-tls-tls13.md]: And while we’re out it, yes, let’s just make it an integer so it maps to whatever x.y or x or /x we use in the future.

File: draft-ietf-tls-tls13.md (hunk 2 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: Done.

File: draft-ietf-tls-tls13.md (hunk 3 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: Done.

**Comment by ekr** [comment]
Separately merged

**Pull request closed without merging**